### ISSUE TITLE
Improper conversions of object dtype to string[pyarrow]

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#10631]
I'm really happy about the use of pyarrow string by default for strings.  I think that it solves a lot of performance problems that made things unusable.  

However, it seems like there are also frequent user errors about this feature, probably not enough to warrant turning it off, but enough that we should maybe think about it at a high level.  My sense is that people have various code that creates non-string objects and that we overly eagerly convert those to string[pyarrow] type.  Situations that I've observed:

- types like 
    - decimal
    - list
    - date
    - struct
-  when ...
    - reading parquet
    - specifying `meta=("name", object)`
    - shuffling?

I suspect/hope that there are ways to reduce some of these errant conversions that don't stop the general benefit of using pyarrow strings.  For example, I suspect that we could add mappings to the arrow type mapper when reading parquet.  I did this in https://github.com/dask-contrib/dask-expr/pull/373 and it worked fine I think?  Probably if the user specifies `object` as a dtype then we should respect it.  

I'm curious to learn more about what's possible and not possible when solving this problem.  My guess is that that we can improve the situation, even if we don't solve it entirely.

cc @phofl @jrbourbeau @fjetter

**Comment by fjetter** [comment]
I would like input from @jrbourbeau but IIUC the current implementation doesn't allow us to distinguish much here. 

If this toggle is on, a `map_partitions` is scheduled that will cast the columns. For instance, even if the parquet reader would deal with other dtypes in a better way, a subsequent `map_partitions` would kill it again. I currently don't see an easy way out of this.

Similarly, regarding structs, lists, etc. we don't have a way to know without seeing the data and we rely on this casting when dealing with `meta` already. The cast itself could likely be made a little smarter but we'd have to deal with meta inconsistencies all over the place.

@phofl is there anything in pandas we can push / wait for to make this experience nicer?

**Comment by phofl** [comment]
https://pandas.pydata.org/docs/whatsnew/v2.1.0.html#avoid-numpy-object-dtype-for-strings-by-default

This is the best we can do, it converts the data through looking at them. It would cause meta and the actual data get out of sync, but that's the only downside I can see

**Comment by mrocklin** [comment]
Just to be sure, what I'm saying is "there are probably some ways to improve things, even if we don't entirely solve the problem".  

As an example, I found that `date` types were converted to object and then to string[pyarrow].  By convering `date` instead to timestamp / datetime64 in the parquet reader they turned into a more reasonable data type and it seemed like they stayed that data type in subsequent operations (I doubt that the code from James converts a datetime64 to a string[pyarrow]).  

Is my understanding above correct?  If so then great! let's spend a little bit of effort on type mappers in the parquet reader.  I'll bet that that helps reduce the pain here.  Maybe there are other low-hanging fruit like this?

**Comment by fjetter** [comment]
Yes, the `date` type should be better handled this way. I was thinking mostly about the lists and more complex things.

**Comment by mrocklin** [comment]
Cool.   What other low hanging fruit exist?  Decimal?  

Additionally, what about another approach?  Right now it looks like we do this in `Frame` so it's very global.  Maybe if we make all of the data ingestion functions better (parquet, csv, snowflake, ...) then we don't need to do as eager of a conversion and we let object dtype stay object dtypes?  I suspect that most of the benefit here is handled at the IO layer.

**Comment by phofl** [comment]
So I looked a little into the types mapper, converting dates and decimals to numpy dtypes is not supported out of the box. We could theoretically cast them ourselves if we want to?

**Comment by rjzamora** [comment]
Just a note that we are running into this problem a lot in RAPIDS while working on the pandas 2+ support.

Since cudf supports optimized `struct` and `list` dtypes, we have test coverage for the pandas-backed version of struct/list-column processing. For these kinds of workflows, it is obviously a bug to convert the column to `string[pyarrow]`, because the processing logic is not expecting a string.

For now, we are just setting the `"dataframe.convert-string"` config to `False` in a bunch of tests, because I do realize that this is difficult to "fix" (and encouraging pyarrow strings by default still seems like a good idea).
